ui/pages/permissions-connect/permissions-connect.component.js

```diff
@@ -175,8 +177,8 @@ export default class PermissionConnect extends Component {
   }
 
   componentDidUpdate(prevProps) {
-    const { permissionsRequest, lastConnectedInfo, origin } = this.props;
-    const { redirecting } = this.state;
+    const { permissionsRequest, lastConnectedInfo } = this.props;
+    const { redirecting, origin } = this.state;
```

Comment:
```suggestion
    const { permissionsRequest, lastConnectedInfo, targetSubjectMetadata } = this.props;
    const { redirecting, origin } = this.state;
    
    if (targetSubjectMetadata && targetSubjectMetadata.origin) {
      this.setState({ targetSubjectMetadata });
    }
```

What if we do something like this?

Reply:
I was thinking that as well, but that change reintroduces the display of "unknown origin" :thinking:

Reply:
maybe it is that change plus something else...

Reply:
Ah. You're right. So this caching breaks the Snaps connection flow 🤔

Reply:
I think this works:
```
    const { permissionsRequest, lastConnectedInfo, targetSubjectMetadata } = this.props;
    const { redirecting, origin } = this.state;

    if (prevProps.targetSubjectMetadata?.origin !== targetSubjectMetadata?.origin && 
         targetSubjectMetadata?.origin) {
      this.setState({ targetSubjectMetadata });
    }
```

I tested it and the unknown origin bug and #29594 are fixed. Just need  to confirm that the snaps case is still working as well

Reply:
Yep, I was just updating my suggestion to a similar thing as well

Reply:
I believe that the headers on snaps connect request are still working as well

Reply:
Yeah, from my testing the snippet above seems to do the trick

Reply:
Updated here https://github.com/MetaMask/metamask-extension/pull/29880/commits/56c81e263c0c502d9349b4a94ab21aa847b30ac2